girder/api/rest.py

```diff
@@ -73,6 +75,36 @@ def getApiUrl(url=None):
     return url[:idx + 7]
 
 
+def iterBody(length=READ_BUFFER_LEN):
+    """
+    This is a generator that will read the request body a chunk at a time and
+    yield each chunk, abstracting details of the underlying HTTP server. Data
+    is returned as soon as it is available, so it may be less than the specified
+    buffer length.
+
+    This works regardless of whether the body was sent with a Content-Length
+    or using Transfer-Encoding: chunked.
+
+    :param length: Max buffer size to read per iteration.
+    :type length: int
+    """
+    if cherrypy.request.headers.get('Transfer-Encoding') == 'chunked':
+        cherrypy.request.rfile.bufsize = length
+        while True:
+            cherrypy.request.rfile._fetch()
+            if cherrypy.request.rfile.closed:
+                break
+            buf = cherrypy.request.rfile.buffer
+            cherrypy.request.rfile.buffer = ''
+            yield buf
+    elif 'Content-Length' in cherrypy.request.headers:
+        while True:
```

Comment:
I don't have strong feelings either way, though I'm also not fond of seeing `cherrypy.request.body.read(length)` duplicated.